### ISSUE TITLE
Typescript autocomplete crash over time

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/TypeScript#8717]
_From @MLefebvreICO on May 18, 2016 15:35_

- VSCode Version: 1.1.1
- OS Version: Windows 10 Pro x64

Steps to Reproduce:

1. Create a  `TypeScript` project with the following structure : 
`tsconfig.json`
`app/`
`app/application.ts`
`app/framework/`
`app/framework/application.ts (included in application.ts)`
`app/framework/some other typescripts files (included in framework/application.ts)`
`app/appNameHere/`
`app/appNameHere/application.ts (included in application.ts)`
`app/appNameHere/some other typescripts files (included in appNameHere/application.ts)`

 So basically : 
 - `app/application.ts` will only contain references to `framework/application.ts` and `appNameHere/application.ts`.

 - `app/framework/application.ts` will only contain references to files from `framework` folder.
 - `app/appNameHere/application.ts` will only contain references to files from `appNameHere` folder.

2. Create a `tsconfig.json` as follow : 
`{
    "compileOnSave": true,
    "compilerOptions": {
        "target": "es5",
        "removeComments": true,
        "outFile": "application.js",
        "sourceMap": true,
        "noImplicitAny": true
    },
    "files": [
        "app/application.ts"
    ]
}`
3. Add some TS files in both directory and play with it, try using intellisense of a class from another file

4. Up at some point, TypeScript intellisense seems to crash and to get it again, we need to restart vscode

The Intellisense is crashing a lot in one day on my side with that settings at my job, it will say it doesn't know what is a class that should be defined, until you restart vscode ...

If you need more informations, I'll be glad to offer them. However, I am not in the ability to share my projet as it is confidential files from my job.

_Copied from original issue: Microsoft/vscode#6485_

**Comment by dbaeumer** [comment]
_From @MLefebvreICO on May 18, 2016 20:20_

Forgot to mentionned : 
My tsc is version 1.8.9

**Comment by dbaeumer** [comment]
@MLefebvreICO could you please enable TS logging. This can be done by adding the following setting:
```json
{
    "typescript.tsserver.trace": "messages"
}
```
Then open the output window and switch to the TypeScript channel. Could you provide me with the log data printed there. That would help to understand why intellisense stops working

**Comment by dbaeumer** [comment]
_From @MLefebvreICO on May 19, 2016 12:11_

@dbaeumer  I just added it to my `settings.json`file. However, I do not have an option TypeScript in the dropdown of the output window even after restarting vscode. 

Am I missing something here ?

EDIT: 
Note that I do not compile with `tsc` directly, but with a gulp command. That should not interfere with the autocompletion.

**Comment by dbaeumer** [comment]
_From @MLefebvreICO on May 19, 2016 12:25_

After playing with my files and everything, it finally appeared.

For now, here's the output, but it didn't crash in that part. Do you prefer that I send one once it stopped working ?

As for now : http://pastebin.com/raw/DZwegjDX

**Comment by dbaeumer** [comment]
@MLefebvreICO yes, I prefer having a trace when it stopped working since the trace will have information why the request failed.

**Comment by dbaeumer** [comment]
_From @MLefebvreICO on May 19, 2016 20:23_

I have not come across the bug today.
Note that when I did this issue I did check if I had the latest TypeScript version and updated it to `1.8.10`.

Maybe that fixed it ? I'll come back to you tomorrow at the end of the day to tell you if it ocurred or it seems fixed.

`TypeScript` release note : 
> This release fixes an issue in React 15.0 and later by removing a dependency on an internal React API when emitting JSX spread attributes. As a result, TypeScript now emits its own standalone helper function for spread attributes.

**Comment by dbaeumer** [comment]
_From @MLefebvreICO on May 20, 2016 15:34_

@dbaeumer I have come across an error. This is not the errors I normally get (they seems gone with the new typescript version).

The code that got an error on is : 
`this.someAngularFactory.someFunction(param1, param2).then...` 

Giving me this error on mouse hover : `Supplied parameters do not match any signature of call target`. However, the function name is perfectly fine, same for the parameters number and their types.

Restarting vscode fixed the issue ...

Here is the trace : http://pastebin.com/raw/dXqXQdsD

**Comment by dbaeumer** [comment]
_From @MLefebvreICO on May 20, 2016 18:7_

Just go an error I normally got when fetching from another branch on git. Seems like git operations screw up the TypeScript intellisense.

I just lost all my references to `angular` and other classes I use.

The only trace that it gives is : 
`Sending request: quickinfo (568). Response expected: yes. Current queue length: 0`
`Response received: quickinfo (568). Request took 2 ms. Success: false . Message: No content available.`

**Comment by dbaeumer** [comment]
@MLefebvreICO thanks a lot for the trace. The trace shows that the tsserver returns failed responses due to a length < 1 violation. Moving to TS team.

**Comment by dbaeumer** [comment]
I also opened https://github.com/Microsoft/TypeScript/issues/8716 to make tracing this more easy.

**Comment by mhegazy** [comment]
@zhengbli can you take a look.

**Comment by SergioMorchon** [comment]
The dame behavior happened to me daily, multiple times.
I work in a JS project, with some react components, mixing it with some .d.ts to describe more precisely some JS code.
And after some time, IntelliSense stops giving me the autocomplete options, only showing the weird basic options (keywords and already in-document present words).
I think, if it crashes, at least let it know to the user.

**Comment by MLefebvreICO** [comment]
It's been 2 weeks, any news would be appreciated !

**Comment by zhengbli** [comment]
@MLefebvreICO I tried the setup described in the original comment, though didn't get a repro. Can you suggest a more concrete repro steps? My project set up was https://github.com/zhengbli/i8717. Feel free to add more steps if I am missing something. 

Also it is hard to tell what went to wrong with the trace you provided previously (all I know is the server failed at some point). I'm working on returning the trace as Dirk suggested.

FYI we changed how file watching works after 1.8 release, so you can give the nightly a try to see if that fixes anything. To use the nightly with VSCode, you can follow the instruction at https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Nightly%20Builds.md#visual-studio-code

**Comment by MLefebvreICO** [comment]
I will test later on, through I think this happens when we switch branch and there's a lot of changes between the branches.

Will confirm if it still happens with the `Nightly`

**Comment by MLefebvreICO** [comment]
@zhengbli After testing with the Nightly it still happens. Is #8716 in the new Nightly ? If so I should be able to tell your the errors returned by TSServer

**Comment by MLefebvreICO** [comment]
@zhengbli Using Nightly autocomplete stopped working too, but instead of putting errors in red, when I put my mouse over it, it says "Loading..." and stays that way.

TSServer returns this multiple times (with a different sequence number) : 

`TypeScript Service: canceled request with sequence number 1225`


I cannot suggest a more concrete repo steps, since it's like what we have here. The only thing I could say is add a lot of files using classes and change a lot of stuffs using them in the branches and switch branches. Sometimes the autocomplete stops working, sometimes not.

**Comment by zhengbli** [comment]
I think the changing many files all at once might have caused the problem. Though still there are not enough information to tell yet. @MLefebvreICO the trace printing bit is merged into master, so tomorrow's nightly should include the trace printing. Or if you would like to, you can download my locally built version: [local.zip](https://github.com/Microsoft/TypeScript/files/305851/local.zip). After you got that, you can try set the `"typescript.tsserver.trace"` to be `"verbose"`, which gives more information.

**Comment by SergioMorchon** [comment]
@zhengbli after changing the "typescript.tsserver.trace" to be "verbose" in my vscode, where can I find that output?

**Comment by MLefebvreICO** [comment]
@SergioMorchon In `View` -> `Toggle Output` (or `CTRL`+`SHIFT`+`U`) and then select `Typescript` in the dropdown at the right of the panel.

**Comment by SergioMorchon** [comment]
Here it is!:

I've renamed the filename, and also ommited the filelist. Said that, after working a little in a project with about 560 files (exluded by jsconfig, there are 279), I hover the line number 90 column whatever, and the typescript output says:

```
Sending request: projectInfo (1608). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "contact.js",
    "needFileNameList": true
}

Response received: projectInfo (1608). Request took 5 ms. Success: true 
Result: {
    "configFileName": "jsconfig.json",
    "fileNames": [/* 279 files */]
}

Sending request: occurrences (1610). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "contact.js",
    "line": 90,
    "offset": 17
}

Response received: occurrences (1610). Request took 85 ms. Success: true 
Result: [
    {
        "start": {
            "line": 13,
            "offset": 7
        },
        "end": {
            "line": 13,
            "offset": 27
        },
        "file": "contact.js",
        "isWriteAccess": true
    },
    {
        "start": {
            "line": 90,
            "offset": 9
        },
        "end": {
            "line": 90,
            "offset": 29
        },
        "file": "contact.js",
        "isWriteAccess": false
    },
    {
        "start": {
            "line": 93,
            "offset": 5
        },
        "end": {
            "line": 93,
            "offset": 25
        },
        "file": "contact.js",
        "isWriteAccess": false
    }
]

```

**Comment by MLefebvreICO** [comment]
@zhengbli Even with the new Nightly and with the `.zip` you provided, there's not much information returned while having `"typescript.tsserver.trace": "verbose"`.

Still getting : `TypeScript Service: canceled request with sequence number 8315`

**Comment by zhengbli** [comment]
@SergioMorchon Judging from the log, the server is functioning correctly, as all the requests were successfully processed. However if you were hovering on an identifier, VSCode should send out a `quickInfo` request, which doesn't show here. That's unexpected, but the server didn't crash in this case.

**Comment by zhengbli** [comment]
@MLefebvreICO I think the `canceled request` message you saw is what VSCode does when it detected a failed request. So that would happen after the server crashes. If you can catch the very log message when the crash happens it would be even more helpful.

**Comment by MLefebvreICO** [comment]
It doesn't look like there's something before the `canceled request` that fail. But if it can helps : http://pastebin.com/raw/Kx8Ffb3G

If it don't helps then vscode or TSServer is not returning error trace properly ...

**Comment by MLefebvreICO** [comment]
@zhengbli It doesn't look like there's something before the `canceled request` that fail. But if it can helps : http://pastebin.com/raw/Kx8Ffb3G

If it doesn't helps then vscode or TSServer is not returning error trace properly ...

**Comment by zhengbli** [comment]
@dbaeumer Can you provide some insights on when does VSCode decides to cancel requests? Judging from the log, it seems that right before the cancelling tsserver did not crash.

**Comment by dbaeumer** [comment]
@zhengbli since the tsserver is basically single threaded we deliver (except of change notifications) only one request at a time. If VS Code cancels a request in the UI (for example the user presses escape) then we don't deliver the request and the log contains:
```
TypeScript Service: canceled request with sequence number 329
```

If the request got already delivered we can't cancel the request. We print
```
TypeScript Service: tried to cancel request with sequence number 263. But request got already delivered.
```

**Comment by MLefebvreICO** [comment]
@dbaeumer Except I didn't pressed escape when I got all the canceled requests, it just won't load the intelisense .. :/ 

Maybe this is an issue with vscode after all

**Comment by mhegazy** [comment]
@dbaeumer so what is the resolution here?

**Comment by dbaeumer** [comment]
We automatically cancel request after different conditions which could be: time, user continues typing, cursor movement. So canceling a request doesn't mean any form of error. Reading through the issue again I still have the feeling that the tsserver fails on some request hence to result is provided.

Just to clarify: We currently don't show any error in the UI if a request fails. What we do is for example to show an empty code completion window. Agree that we need to improve this as well.

**Comment by dbaeumer** [comment]
Opened https://github.com/Microsoft/vscode/issues/8716

**Comment by SergioMorchon** [comment]
I had another way to reproduce a similar behavior: vscode stops showing the flyout with typescript hint on mouse over.
Steps to reproduce:
1. Open a TS file (I did it with https://github.com/pladaria/reconnecting-websocket/blob/master/index.ts).
2. Hover over any function, to confirm that the intellisense shows the resume of the function. In my case, I hovered the inner `connect` function.
```
Sending request: quickinfo (2). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "/Users/smorchon/workspace/personal/reconnecting-websocket/index.ts",
    "line": 97,
    "offset": 15
}

Response received: quickinfo (2). Request took 18 ms. Success: true 
Result: {
    "kind": "const",
    "kindModifiers": "",
    "start": {
        "line": 97,
        "offset": 11
    },
    "end": {
        "line": 97,
        "offset": 18
    },
    "displayString": "const connect: () => void",
    "documentation": ""
}
```
3. Perform quickly about 10 right clicks. In my Macbook Pro, the contextual pane opened with each right-click appears to lag at each show, and I had to perform some extras left click elsewhere in the window to let vscode show the remaining contextual panels. ![captura de pantalla 2016-07-04 a las 16 26 25](https://cloud.githubusercontent.com/assets/4521712/16563753/19033440-4204-11e6-8438-edadd3fd326c.png)
4. After that lagged right clicks, vscode stops showing any hinting over the file.
```
Sending request: occurrences (3). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "/Users/smorchon/workspace/personal/reconnecting-websocket/index.ts",
    "line": 97,
    "offset": 14
}

Response received: occurrences (3). Request took 21 ms. Success: true 
Result: [
    {
        "start": {
            "line": 97,
            "offset": 11
        },
        "end": {
            "line": 97,
            "offset": 18
        },
        "file": "/Users/smorchon/workspace/personal/reconnecting-websocket/index.ts",
        "isWriteAccess": true
    },
    {
        "start": {
            "line": 140,
            "offset": 28
        },
        "end": {
            "line": 140,
            "offset": 35
        },
        "file": "/Users/smorchon/workspace/personal/reconnecting-websocket/index.ts",
        "isWriteAccess": false
    },
    {
        "start": {
            "line": 148,
            "offset": 5
        },
        "end": {
            "line": 148,
            "offset": 12
        },
        "file": "/Users/smorchon/workspace/personal/reconnecting-websocket/index.ts",
        "isWriteAccess": false
    }
]
```

As you can see, the first time tsserver returns a displayString, but the second time does not!

**Comment by MLefebvreICO** [comment]
With Typescript `1.9.0-dev.20160609-1.0`, I just got some errors while changing branches (where normally it would say nothing and make the intellisense stop working).

Here are some errors : 
* `Module 'Innova' has no exported member 'IAttachmentFileOrFolder'`
* `Cannot find name 'IAttachmentFile'` (which is on our `Innova` module)
* `Cannot find name 'angular'`

When changing branches TypeScript does not seems to reload the tsconfig properly as these definitions are provived in there (See first post to see how we set it).

**Comment by MLefebvreICO** [comment]
@mhegazy In fact, I don't know how TypeScript watch for files, but if it's like our `gulp watch` task, it crashes when there was a folder watch that suddendly doesn't exists anymore.

This is an error with `FSWatcher` while using `Gaze` lower then `1.0`, see : https://github.com/gulpjs/gulp/issues/651

**Comment by nullptr128** [comment]
I'd also point out that I've same issue in Atom with atom-typescript plugin so it is not only VSCode issue.

**Comment by MLefebvreICO** [comment]
@nullptr128 Are you also on Windows ? If so that may confirm my idea of the bug.

**Comment by nullptr128** [comment]
No I am on Linux Mint 17 (Ubuntu 14.04) but the issue is the same. Atom completes code correctly up to a some point where it just stops doing so. Only restarting IDE works. AFAIK typescript code completion in various IDEs is supported by built-in typescript autocomplete daemon so I believe the issue is same here.

Not sure how I could turn on logging these crashes in Atom, if you'd point me how to do it, I could also send my stack dumps when it crashes.

**Comment by MLefebvreICO** [comment]
@mhegazy Did you have time to check that out a bit ? I'm really getting tired of rebooting my IDE every time I change a branch because of TypeScript crashing.

**Comment by zhengbli** [comment]
@MLefebvreICO not much progress has been made as I still have problems reproing this issue. I tried deleting a folder (with `/// reference` in another file pointing to a file in this folder), and switch branches, the server worked as expected.

@nullptr128 can you provide a sample setup that reproduces the issue? That would help the investigation greatly. Thanks!

**Comment by MLefebvreICO** [comment]
@zhengbli May I ask on what OS ?

**Comment by zhengbli** [comment]
I'm on Windows 10 64-bit, with the latest insider builder of VSCode (1.4.0-insiders) and TypeScript nightly (`2.1.0-dev.20160713`)

**Comment by MLefebvreICO** [comment]
Well, in the meantime, I have found that we can do keyboard shortcut to reload the typescript project in vscode by adding this to the keyboard shortcuts .. so no more complete reload of vscode anymore.

``` json
{
        "key": "ctrl+shift+f5",
        "command": "typescript.reloadProjects"
}
````

Hopefuly this can help some more people while this is getting fixed.

But @zhengbli, what can I possibly do to help fix this ?

Since TSC doesn't return an error when it crashes ...

**Comment by MLefebvreICO** [comment]
@zhengbli @dbaeumer Is there a way that I could connect directly to the TypeScript language server launched by vscode to see the error directly ?

**Comment by MLefebvreICO** [comment]
Putting back the logs to be verbose.

It crashed and `angular` was not found by TypeScript, vscode added a red underline to it.

TypeScript verbose logs : 

``` json
Sending request: quickinfo (1863). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "c:/ico/BeeON.Prototype/Website/innova/app/directives/attachments/attachments.controller.ts",
    "line": 237,
    "offset": 27
}

Response received: quickinfo (1863). Request took 20 ms. Success: true 
Result: {
    "kind": "",
    "kindModifiers": "",
    "start": {
        "line": 237,
        "offset": 25
    },
    "end": {
        "line": 237,
        "offset": 32
    },
    "displayString": "any",
    "documentation": ""
}
```` 

So as said in one of my comments, it looks like TypeScript didn't reload all the typings files properly since `angular.d.ts` is in fact included. 

Couldn't you just listen to git events then when we change branches, the TypeScript reload all the typings ?!

**Comment by zhengbli** [comment]
In your case, the server did not crash (it still resolved the `quickInfo` request successfullly, though incorrectly), therefore it doesn't have errors to send. Here is what I tried according to your description:
![gif](https://cloud.githubusercontent.com/assets/1171301/17257264/25c42b3c-5576-11e6-9044-c224c6b29daf.gif)

Tsserver only has access to the underlying file system therefore can only react to file system events. Maybe VSCode can react to git events (I don't know if they already do that).

Really if a stable repro setup can be found on this, the progress can be made much faster. Unfortunately, all the steps / procedures provided in the comments here didn't do that (Or if you find a stable repro requires other OS / environment variables, let me know as well). My best understanding according to these piece-meal information in the post is that in some cases when switching git branches tsserver didn't include all project files but it didn't crash.

**Comment by MLefebvreICO** [comment]
@zhengbli Even on my setup it doesn't happen all the time. And like I said, I cannot share it since it's work.

Already I can see two things that differ on your setup : 
* You didn't reproduce the correct tree. Having an `application.ts` file that has all the references and other files that does not.
* We use classes inside of namespace, so maybe creating two modules and having one of these two refering to a function in the class of the other namespace.

Also like a mentionned it looks like it has more chance of happening when you go from a branch that had a reference on a file inside a folder that doesn't exist on the other branch.

I'll try to setup a sample project this weekend and try to reproduce it.
I hope I will, because this is becoming really frustating when it happens sometimes about 10 times a day.

**Comment by MLefebvreICO** [comment]
See my [test repo](https://github.com/MLefebvreICO/typescript-8717)

To see the error in action, simply open vscode insiders go on the file `b/some-files/b1.ts` and play between branches `master` and `test`. The `a.A1` is moved to another file in `test` branch and is now `a.A2`.

TypeScript doesn't seems to update his references, because when I switch from `test` to `master`, I'm getting : 

> Property 'A1' does not exist on 'typeof a'

**Comment by nullptr128** [comment]
Not sure if this is related at all, but in Atom it looks like it fixed.....itself? I do not really know what happened here, but it seems to work correctly now on my all projects. I think I have updated atom-typescript package to latest version recently so probably my issue is not related at all.

**Comment by MLefebvreICO** [comment]
@nullptr128 If your issue is not related, then I'm pretty sure it has to do with how TypeScript watch files on Windows as previously mentionned in one of my message

**Comment by zhengbli** [comment]
@MLefebvreICO Thanks for the repro sample! Now I can get the issue as well. Investigating.

**Comment by MLefebvreICO** [comment]
@zhengbli Great ! Let me keep updated !

**Comment by zhengbli** [comment]
So this happens because the editor doesn't know when do git operations finishes, two situations could happen:
1. The current open file (b) changed **before** the file it referenced (a) changed, then when the editor asks the server for errors in the current open file (b), the copy of file a is the old version. As a result, you see an outdated error.

2. The current open file (b) changed **after** the file it referenced (a) changed, then when the editor asks the server for errors in the current open file (b), the copy of file a is the new version. Things are good.

Because either case could happen, I do see the occasional appearance of this issue. However, if I make any edit in the current open file after the git operation, the error goes away. Do you see the same thing?

**Comment by zhengbli** [comment]
@MLefebvreICO even in your repro sample editing doesn't make a difference? That's different from what I am seeing. Would you be available sometime this week or early next week for a skype session? I feel it would be more efficient if we can just chat for a bit. My email is zhengbli@microsoft.com if needed.

**Comment by MLefebvreICO** [comment]
@zhengbli Actually you were right. I had forgotten to change the `typescript.tsdk` path in my vscode insiders so it was using the TypeScript version included in vscode.

Both cases as you described them are happening here too.

The problem now, as is, seems to be less relevent since, like you mentionned, we can simply edit the file and the typings gets updated.

Also, after testing with TypeScript 2.0, it works the same. So my only left concern is to know when vscode will update it's TypeScript to 2.0, but I'll see with them for that.

As for myself, you can consider this issue resolved.

**Comment by zhengbli** [comment]
As we have completely re-implemented tsserver for 2.0, and I still cannot repro this issue (the closest thing I got is an incorrect error showing up right after git branch switch; but it will go away after any edit on this file, which IMO is acceptable as there is no way for us to know when does a git operation finish). Please reopen the issue with a repro step if you still see it happening on 2.0. Thanks.
